bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicationWorker.java

```diff
@@ -364,40 +368,52 @@ private void deferLedgerLockRelease(final long ledgerId) {
         TimerTask timerTask = new TimerTask() {
             @Override
             public void run() {
+                boolean isRecoveryOpen = false;
                 LedgerHandle lh = null;
                 try {
                     lh = admin.openLedgerNoRecovery(ledgerId);
                     if (isLastSegmentOpenAndMissingBookies(lh)) {
+                        // Need recovery open, close the old ledger handle.
+                        lh.close();
+                        // Recovery open could result in client write failure.
+                        LOG.warn("Missing bookie(s) from last segment. Opening Ledger{} for Recovery.", ledgerId);
                         lh = admin.openLedger(ledgerId);
+                        isRecoveryOpen = true;
                     }
-
-                    Set<LedgerFragment> fragments = getUnderreplicatedFragments(lh);
-                    for (LedgerFragment fragment : fragments) {
-                        if (!fragment.isClosed()) {
-                            lh = admin.openLedger(ledgerId);
-                            break;
+                    if (!isRecoveryOpen){
```

Comment:
hmm, it seems this logic is changed, no?

Reply:
@sijie  All we are doing in this block is to open the ledger in recovery mode isn't it?

Reply:
@jvrao I see.

to make things a bit clear here, it might be good to move line 384 to 395 to a function call "hasOpenUnderreplicatedFragements(LedgerHandle lh)`.

and change the if else here to be:

```
if (isLastSegmentOpenAndMissingBookies(lh) || hasOpenUnderreplicatedFragments(lh)) {
      lh.close();
      lh = admin.openLedger(ledgerId);
      isRecoveryOpen = true;
}
```

not a blocker though.